### ISSUE TITLE
usecase layer of clean achitecture for creating a new game

### ISSUE THREAD

**Issue by metalalive** [Game-as-a-Service/Pandemic-Reign-of-Cthulhu#10]
Discuss the usecase (application scenario) , for the endpoint in current [Open API doc](https://game-as-a-service.github.io/Pandemic-Reign-of-Cthulhu/doc/api)  :
- `/games` in `POST` method

### This comment will be updated irregularly, and collecting opinions and discussions from current backend members.

討論 創建新遊戲的 usecase layer 該做什麼 ?  此評論會不時更新

When creating a new game, usecase should do the followings :
- validate whether number of players does exceed the limit (min. 1, max. 4 players) 驗證玩家數量是否超過限制
- initially , randomly assign investigator cards to current players in the new game 在新遊戲中,將調查員卡隨機分配給當前玩家
  - remind that any user can switch to different investigator card at **a later time before starting the game** 用戶可以在**開始遊戲之前**切換到不同的調查員卡
  - another API endpoint doing this is `/games/{game-id}/investigator` with POST method
- default difficulty level (introductory ? or standard ?) 預設難度等級
  - conclusion : **introductory level** by default

Technical Detail :
- implement the flow above
- declare repository named `InMemoryRepository`  for storing state of all ongoing games
  - for now the repository can be in-memory data store, [because database decision can be deferred](https://www.linkedin.com/pulse/clean-architecture-jo%C3%A3o-h%C3%A9lio-%C3%A1vila-miranda).
  - current interface of  `InMemoryRepository`  : `save()`, other interface methods will be added later in another GitHub issue.
- declare domain model named `Game` and `Player`
  - its data fields and methods detail will be added later in another GitHub issue.

**Comment by metalalive** [comment]
@teds-lin , @wingtkw
please leave your comment at here for how we can implement this,
thanks

**Comment by metalalive** [comment]
@wingtkw 
> better to be introductory level

Sure

> aren't we allowing the users to choose between 2 characters cards

I am reading the rule book, I guess what you mean by `character cards` is `investigator cards`.

Let me clarify this. When I described : 
> randomly assign investigator cards to current players 

- This is only for initializing everything when creating a game , in case that the users straightly start the game without selecting investigator cards. However it is optional (that's why we need discussion)
- Users can still switch to any investigator card that hasn't been chosen yet (and discard the previously selected one) **a later time**,  **in different API endpoint** (POST method, `/games/{game-id}/investigator`) , this should be discussed later in another issue

**Comment by metalalive** [comment]
(correct some grammar mistakes)
@wingtkw 

> better to be introductory level by default

Sure

> aren't we allowing the users to choose between 2 characters cards

I am reading the rule book, I guess what you mean by `character cards` is `investigator cards`.

Let me clarify this. When I described : 
> randomly assign investigator cards to current players 

- This is only for initializing everything when creating a game 
  - In case that the users start the game immediately after creating a new game  without selecting investigator cards.
  - It can be optional , we can also do this when the users are starting the game.

- Users can still switch to any investigator card that hasn't been chosen yet, and discard the previously selected one **at later time**, and  **using different API endpoint** (POST method, `/games/{game-id}/investigator`)
  -  This should be discussed later in another github issue
  - the API endpoint  `/games/{game-id}/investigator` comes from previous meeting and recorded in [Miro Page](https://miro.com/app/board/uXjVMO8BrcQ=/)

**Comment by metalalive** [comment]
Other feedback and questions relevant to this GitHub issue are welcome, thanks.

If there is nobody replied by 13:00, 27th. Oct (GMT+8) , I'll start the implementation of the use case layer.

**Comment by teds-lin** [comment]
參考Miro，玩家可選擇不同難度
- 教學難度
- 基礎難度
- 專家難度

introductory 對應教學而 standard 對應基礎嗎？
![image](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/assets/125881849/137fa6a1-4e32-4490-b0c7-89f76c8ba9c0)

**Comment by metalalive** [comment]
> introductory 對應教學, 而 standard 對應基礎嗎

是的，你可以這樣想

但是我正在考慮將 standard 變更中文命名為 "**標準難度**", 
- 一方面是它符合字典的定義, 
- 另一方面玩家或許對此感到困惑  -->  "教學" vs. "基礎" , 哪一個是最簡單的級別 lol

**Comment by teds-lin** [comment]
> see [current implementation code](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/blob/main/backend/app/usecase/config_game.py)

fix link
 [current implementation code](https://github.com/Game-as-a-Service/Pandemic-Reign-of-Cthulhu/blob/fc119e2a111c8c1431e3020120d787ed427b1b68/backend/app/usecase/config_game.py)
